### DISCUSSION TITLE
Mapping from polyRef -> tilePolyIndex

### DISCUSSION THREAD

**Discussion by rob-myers** [isaac-mason/recast-navigation-js#307] [General]
Given a polygon reference `polyRef` (a number) we can visualise it using `tile.detailMeshes(tilePolyIndex)` as in [`getDebugNavMesh`](https://github.com/isaac-mason/recast-navigation-js/blob/d3adca721ce1993976d4cc01881a61e76e9ecec3/packages/recast-navigation-core/src/nav-mesh.ts#L429).

In particular:
- `polyRef` is global over the entire navMesh
- `0 ≤ tilePolyIndex < tile.polyCount()`.

In order to map between them I am currently building a lookup from `poly.raw.ptr` to each respective `tilePolyIndex`. I also have to update this lookup whenever a tile changes e.g. add/remove obstacle.

Is there a better way?

**Comment by rob-myers** [comment]
In this case we could expose [`dtNavMesh::decodePolyId`](https://recastnav.com/classdtNavMesh.html#afaaf5c22864cdf4a61d5e40a4efce43f).

We could also provide its inverse i.e. `dtNavMesh::encodePolyId`.

**Reply by isaac-mason** [reply]
That sounds sensible to me. I'll create an issue for this. Feel free to create a PR if you need it soon, otherwise I'll get to it at some point.

**Reply by isaac-mason** [reply]
https://github.com/isaac-mason/recast-navigation-js/issues/314

**Reply by rob-myers** [reply]
Sure, I will implement this.

**Reply by isaac-mason** [reply]
Much appreciated! 🙏

**Reply by rob-myers** [reply]
See https://github.com/isaac-mason/recast-navigation-js/pull/325.

**Reply by rob-myers** [reply]
@isaac-mason thanks for merging and also your improvements.

I noticed the intersecting triangles don't always seem correct e.g for queryPolygons. The code looks like it's checking for intersecting aabbs. But sometimes it looks like the cube does not intersect these bounds.

**Reply by isaac-mason** [reply]
No problems @rob-myers!

Are you seeing this issue on the "Click Nearby Polygons" storybook?

**Reply by rob-myers** [reply]
Yeah @isaac-mason , the "Click Nearby Polygons" story.

But after another look I'm probably misunderstanding the "detail mesh".

<img width="588" alt="image" src="https://github.com/isaac-mason/recast-navigation-js/assets/1658537/b073e377-325d-4c4e-8ecd-2f39a8a3a789">

The AABB of the triangle containing `322` does not intersect the cube.
However, `322` actually corresponds to this triangle:

<img width="641" alt="image" src="https://github.com/isaac-mason/recast-navigation-js/assets/1658537/8f0094f5-be17-448c-8ccb-c35bd289ae0a">

So presumably I was looking at a "detail triangle", without realising it.

**Reply by isaac-mason** [reply]
Somewhat related - I've been meaning to improve the threejs visualisation helpers for a while - https://github.com/isaac-mason/recast-navigation-js/issues/165

I'm going to put some more energy into this soon. It should help with investigating and understanding things like this.

**Reply by isaac-mason** [reply]
fyi @rob-myers - added a DebugDrawer class to @recast-navigation/three

https://recast-navigation-js.isaacmason.com/?path=/story/debug-three-debug-drawer--poly-mesh
https://recast-navigation-js.isaacmason.com/?path=/story/debug-three-debug-drawer--poly-mesh-detail

**Reply by rob-myers** [reply]
@isaac-mason great job on all those examples!
